src/Agda2Hs/Compile/Term.hs

```diff
@@ -20,311 +21,446 @@ import Agda.Syntax.Internal
-    compileCon h i es
-      | Just semantics <- isSpecialCon (conName h)
-      = semantics h i es
-    compileCon h i es =
-      isUnboxConstructor (conName h) >>= \case
-        Just _  -> compileErasedApp es
-        Nothing -> (`app` es) . Hs.Con () =<< compileQName (conName h)
-
--- `compileErasedApp` compiles an application of an unboxed constructor
--- or unboxed projection or transparent function.
--- Precondition is that at most one elim is preserved.
-compileErasedApp :: Elims -> C (Hs.Exp ())
-compileErasedApp es = do
+
+-- * Compilation of constructors
+
+type ConCompileRule = Type -> [Term] -> C (Hs.Exp ())
+
+-- | Custom compilation rules for special constructors.
+isSpecialCon :: QName -> Maybe ConCompileRule
+isSpecialCon = prettyShow >>> \case
+  "Haskell.Prim.Tuple._,_"          -> Just tupleTerm
+  "Haskell.Prim.Tuple._×_×_._,_,_"  -> Just tupleTerm
+  "Haskell.Extra.Erase.Erased"      -> Just erasedTerm
+  "Haskell.Extra.Delay.Delay.now"   -> Just compileErasedApp
+  "Haskell.Extra.Delay.Delay.later" -> Just compileErasedApp
+  _                                 -> Nothing
+
+tupleTerm :: ConCompileRule
+tupleTerm = compileApp' (Hs.Tuple () Hs.Boxed)
+
+erasedTerm :: ConCompileRule
+erasedTerm _ _ = pure (Hs.Tuple () Hs.Boxed [])
+
+
+-- | @compileErasedApp@ compiles the application of unboxed constructors
+-- and transparent functions.
+-- Precondition: at most one argument is preserved.
+compileErasedApp :: Type -> [Term] -> C (Hs.Exp ())
+compileErasedApp ty args = do
   reportSDoc "agda2hs.compile.term" 12 $ text "Compiling application of transparent function or erased unboxed constructor"
-  compileElims es >>= \case
+  reportSDoc "agda2hs.compile.term" 12 $ text "Args" <+> prettyTCM args
+  reportSDoc "agda2hs.compile.term" 12 $ text "Type" <+> prettyTCM ty
+  compileArgs ty args >>= \case
     []  -> return $ hsVar "id"
     [v] -> return v
     _   -> __IMPOSSIBLE__
 
+
+compileCon :: ConHead -> ConInfo -> Type -> [Term] -> C (Hs.Exp ())
+compileCon h i ty args
+  | Just semantics <- isSpecialCon (conName h)
+  = semantics ty args
+compileCon h i ty args = do
+  isUnboxConstructor (conName h) >>= \case
+    Just _  -> compileErasedApp ty args
+    Nothing -> do
+      info <- getConstInfo (conName h)
+      -- the constructor may be a copy introduced by module application,
+      -- therefore we need to find the original constructor
+      if defCopy info then
+        let Constructor{conSrcCon = ch'} = theDef info in
+        compileCon ch' i ty args
+      else do
+        con <- Hs.Con () <$> compileQName (conName h)
+        compileApp con ty args
+
+
+-- * Term compilation
+
+compileTerm :: Type -> Term -> C (Hs.Exp ())
+compileTerm ty v = do
+
+  reportSDoc "agda2hs.compile" 7  $ text "compiling term:" <+> prettyTCM v
+  reportSDoc "agda2hs.compile" 27 $ text "compiling term:" <+> pure (P.pretty $ unSpine1 v)
+
+  case compileSpined ty v of
+    Just cont -> cont
+    Nothing   -> case unSpine1 v of
```

Comment:
Is `unSpine1` here still necessary? All cases with a spine are already handled by `compileSpined`.